### PR TITLE
Add a Find method...

### PR DESCRIPTION
**Pull request by zeroed**
Find by keys.

This `find` methods wants to be as flexible as possible.
Unless a block is given, it returns an `Enumerator`, otherwise it yields
to the block an instance of the found Cassandra entries.

If no keys as passed as parameter, the methods returns (an Enumerator for)
all the results in the table.

If one or more keys are given, the **order matters**; since this methods
will craft a CSQL query, the first key **must** be the `partition key`,
while the followings are intended as the `clustering keys`... that can be
more than one.

If the keys passed as parameter are more than the defined by the table the
query is not valid, it cannot be executed and an error is raised.

May I ask your opinion @stefanofontanelli and @gbagnoli ?
Also about the test...

Thanks!

### CODE REVIEW

**Comment by stefanofontanelli** [comment]
@zeroed what is the difference with Model#get?

File: lib/corm/model.rb [hunk 1]
```diff
@@ -138,6 +140,51 @@ def self.count
...
+    # will craft a CSQL query, the first key **must** be the `partition key`,
+    # while the followings are intended as the `clustering keys`... that can be
+    # more than one.
+    #
+    # If the keys passed as parameter are more than the defined by the table the
+    # query is not valid, it cannot be executed and an error is raised.
+    def self.find(*key_values, &block)
+      return to_enum(:find, *key_values) unless block_given?
+
+      find_key = "find_#{Digest::MD5.hexdigest(key_values.inspect)}"
```
Comment (gbagnoli) [inline lib/corm/model.rb]: @zeroed statement should be prepared on key names not values. Values are the passed to the prepared statements via data binding.

So basically each key _name_ foe which the value is not `nil` should be used as the hash key. Using value will defeat the purpose of having prepared statements

Not sure if clear enough :)

**Comment by gbagnoli** [comment]
@stefanofontanelli this is different from get as it returns multiple results, as it is possibile in cassandra to select 
- without any key
- using only the partition key, and no clustering key(s)
- using partition key and one or more -in order- clustering keys

All this is not supported by get

**Comment by gbagnoli** [comment]
@stefanofontanelli this to select multiple rows omitting some (or all)
clustering keys.

File: lib/corm/model.rb (hunk 1 above)
Comment (zeroed) [inline lib/corm/model.rb]: Perfectly clear... TBH that line was quite a bit a "_sudden gut reaction_" after a long day ... I'll fix it!
Thanks!

:sweat_smile:

**Comment by zeroed** [comment]
I couldn't have said it better... thank you so much @gbagnoli .
This feature is related to a specific need that probably could be highlighted with a better test (just in case) ... furthermore I would add a couple of corner case among the tests...

**Comment by zeroed** [comment]
INVHO, that's it.

:circus_tent: 

Any more hint @stefanofontanelli @gbagnoli ?

File: test/test_utils.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  end
+
+  MODELS = [FakeModel, FakeMultiKeyModel]
+
+  def setup_corm!
+    @logger = Logger.new(STDOUT).tap { |logger| logger.level = Logger::INFO }
+
+    MODELS.each do |model|
+      model.configure(hosts: ['127.0.0.1'], logger: @logger)
+      model.keyspace! rescue nil
```
Comment (gbagnoli) [inline test/test_utils.rb]: `if_not_exists: true` could be set for `keyspace!` as well instead of rescuing nil

File: test/test_utils.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @some_random_keys = %w[ foo bar lol meh ]
+  end
+
+  def teardown_corm!
+    MODELS.each do |model_class|
+      tablename = [ model_class.keyspace, model_class.table ].compact.join('.')
+      model_class.cluster.connect.tap do |connection|
+        connection.execute("TRUNCATE #{tablename};") rescue nil
```
Comment (gbagnoli) [inline test/test_utils.rb:150]: `truncate!` should be used here instead of explicit query

**Comment by gbagnoli** [comment]
@zeroed you could add a `drop_table!` method and use it in the teardown, and remove the `drop keyspace` entirely from the teardown. I'm not sure we want a `drop_keyspace!` method, and for the sake of tests `truncate!` and `drop_table!` would be enough IMO so for now I'd say do not add it.

thoughts?

File: test/test_utils.rb (hunk 2 above)
Comment (zeroed) [inline test/test_utils.rb]: I agree; thanks.

**Comment by zeroed** [comment]
I totally agree with the suggestions pointed out by @gbagnoli , hence these additional commits.

TBH I has some issue to test the `drop_table!` method... I noticed some _inconsistent_ behaviour between some `puts` and `pry`... since "in pry I trust" I'm quite confident in saying that it works...

Any further feedback is always appreciated.

File: lib/corm/model.rb [hunk 4]
```diff
@@ -138,6 +140,43 @@ def self.count
...
+    # all the results in the table.
+    #
+    # If one or more keys are given, the **order matters**; since this methods
+    # will craft a CSQL query, the first key **must** be the `partition key`,
+    # while the followings are intended as the `clustering keys`... that can be
+    # more than one.
+    #
+    # If the keys passed as parameter are more than the defined by the table the
+    # query is not valid, it cannot be executed and an error is raised.
+    def self.find(*key_values, &block)
```
Comment (stefanofontanelli) [inline lib/corm/model.rb]: @zeroed I don't like the *key_values argument. I kindly suggest you to use an hash of keys/values because it's not clear from the code what is the order of the values in the array. You have to check primary_key() statement in the class definition, but the worst part is that the order is implicit in the implementation: it's really easy to mess things up with a refactoring (your code cannot cover how other users are using the lib).

File: lib/corm/model.rb [hunk 5]
```diff
@@ -138,6 +140,43 @@ def self.count
...
+      return to_enum(:find, *key_values) unless block_given?
+
+      field_names = []
+      Array(key_values).each_with_index do |value, index|
+        field_names << "#{primary_key.flatten[index]} = ?"
+      end
+
+      statement_find_key = "find_#{Digest::MD5.hexdigest(field_names.inspect)}"
```
Comment (stefanofontanelli) [inline lib/corm/model.rb]: @zeroed don't hash with md5: if we use something like `"find_#{values}"`, where `values` is the hash passed to the method, it will be easy for the user checking the content of `statements` var.
I think it's really hard to understand/debug what is "find_0c88028bf3aa6a6a143ed846f2be1ea4" and from where it's coming from.
"find_{id:'pippo'}" is much better than the hash, you can also use only keys and values; something like "find_id_pippo" or "find_id_pippo_id2_pluto".

**Comment by gbagnoli** [comment]
@stefanofontanelli the order IS important, you can't pass them in what order you prefer, so the *key makes a lot of sense to me. An hash does not make sense as you'll rely on the implicit ordering of how you enter the keys (ruby1.9)

If, however, we want 'keywords' we could always define the method with the signature since we know the exact number and names of the keys

File: lib/corm/model.rb [hunk 6]
```diff
@@ -138,6 +140,43 @@ def self.count
...
+    #
+    # If one or more keys are given, the **order matters**; since this methods
+    # will craft a CSQL query, the first key **must** be the `partition key`,
+    # while the followings are intended as the `clustering keys`... that can be
+    # more than one.
+    #
+    # If the keys passed as parameter are more than the defined by the table the
+    # query is not valid, it cannot be executed and an error is raised.
+    def self.find(*key_values, &block)
+      return to_enum(:find, *key_values) unless block_given?
```
Comment (stefanofontanelli) [inline lib/corm/model.rb]: @zeroed I am pretty sure that the result from cassandra driver is an enumerator. Why do you not just return that enumerator?
I mean, the logic should be: build the query, run the query, return the original enumerator or iterate on the results using the block.
Suggestion: I think we should add an explicit limit parameter into the method signature (default: nil).

File: lib/corm/model.rb [hunk 7]
```diff
@@ -138,6 +140,43 @@ def self.count
...
+        statement = self.the_find_statement_for(key_values, field_names)
+        statements[statement_find_key] = session.prepare(statement)
+      end
+
+      execute(statements[statement_find_key], arguments: key_values).each do |cassandra_record_|
+        block.call(new(_cassandra_record: cassandra_record_))
+      end
+    end
+
     def self.get(relations)
```
Comment (stefanofontanelli) [inline lib/corm/model.rb:207]: @zeroed I think a good refactoring is: move main logic of `get` into `find` method, then change `get` to call `find(...).first` and return a new corm object.

File: lib/corm/model.rb [hunk 8]
```diff
@@ -138,6 +140,43 @@ def self.count
...
       if statements['get'].nil?
         fields = primary_key.flatten.map { |key| "#{key} = ?" }.join ' AND '
```
Comment (stefanofontanelli) [inline lib/corm/model.rb]: Instead of get all the primary keys as fields it's a matter to select the only ones passed via relations.

**Comment by stefanofontanelli** [comment]
@zeroed I made some comments. Let me know your thoughts.

File: lib/corm/model.rb (hunk 4 above)
Comment (zeroed) [inline lib/corm/model.rb]: What I tried to achieve here was to makes easy to handle those tables with many clustering keys... I agree that in this way it's not explicitly enforced the constraint for the primary key (must be present, must be in the first position...) but IMHO it's easy to query for one, two, or any clustering key... I'm going to do something better with an hash and than I'll show you.

Thanks!

File: lib/corm/model.rb (hunk 7 above)
Comment (zeroed) [inline lib/corm/model.rb:207]: I totally agree...

File: lib/corm/model.rb (hunk 5 above)
Comment (zeroed) [inline lib/corm/model.rb]: Good point... I agree.

File: lib/corm/model.rb (hunk 4 above)
Comment (gbagnoli) [inline lib/corm/model.rb]: IMO an hash is the wrong signature here.

The keys cannot be passed in any order, they _have_ to be in order, and this is a requirement from cassandra.

[`get`](https://github.com/stefanofontanelli/corm/blob/master/lib/corm/model.rb#L145) has a similar signature to this one: a list of keys. Hash are unordered, at least if we don't want to rely on the ordered of insert which I hope it's not what is being proposed :)
The hash makes sense only for the partition keys.

I'm fine with 
- a single parameter, an `Array` of keys
- the current implementation
- a **generated** method. i.e. if the user defined two partition keys and three cluster keys, we can generate a method `def self.find(part_key1, part_key2, clust_key1=nil, clust_key2=nil, clust_key3=nil)`

File: lib/corm/model.rb (hunk 5 above)
Comment (gbagnoli) [inline lib/corm/model.rb]: I'm sorry, am I missing something? **values** are not taken in consideration when generating prepared statements, as they are added when the statements is used via bind.

I agree that avoiding hashes makes the whole thing more explicit, which is good.
So given a partition key `[id, id2]` and clustering keys `[c1, c2]` we end up with three prepared statements:
- `find_id_id2`
- `find_id_id2_c1`
- `find_id_id2_c1_c2`

File: lib/corm/model.rb (hunk 6 above)
Comment (gbagnoli) [inline lib/corm/model.rb]: :+1: for the limit

for the enum: this method is not returning the values from the driver, is supposed to return corm object, so yes, the enum makes sense to me in the current implementation

File: lib/corm/model.rb (hunk 7 above)
Comment (gbagnoli) [inline lib/corm/model.rb:207]: Just be sure to return `nil` if not records are found for `get` (current implementation does that)
`find` should return an empty enumerator instead.

**Comment by zeroed** [comment]
Consider this PR as slightly WIP... 'cause I need only to:
- move the `get` logic into the `find`;
- ensure the `nil` as return of the `get`;
- understand why `Cassandra::Errors::NoHostsAvailable: All attempted hosts failed: 127.0.0.1 (Cassandra::Errors::IOError: Too many open files - socket(2))` 
- have only one `drop_table!` method;

**Comment by zeroed** [comment]
May I ask to try if `Cassandra::Errors::IOError: Too many open files` is only a my issue?

Thanks...

File: lib/corm/model.rb [hunk 9]
```diff
@@ -139,20 +143,74 @@ def self.count
...
+        block.call(new(_cassandra_record: cassandra_record_))
       end
-      cassandra_record_ = execute(statements['get'], arguments: values).first
-      cassandra_record_ ? new(_cassandra_record: cassandra_record_) : nil
+    end
+
+    def self.get(relations)
+      query_options = {
+        limit: 1,
+        statement_key: 'get'
```
Comment (gbagnoli) [inline lib/corm/model.rb:210]: why do we need to set the `statement_key` here? isn't the `get` a `find` with just part keys? shouldn't it be generated correctly at line 188 like for other cases?

File: lib/corm/model.rb [hunk 10]
```diff
@@ -295,5 +365,22 @@ def statements
...
+    # Create and return the proper query to find the C* entries, given an array
+    # of keys.
+    #
+    # @param key_values An array of key names; can be empty, cannot be, in size, greater than the length of the model keys.
+    # @param field_names The "column names" for the `WHERE` clause.
+    def self.the_select_statement_for(key_values, field_names, limit = nil)
+      limit = "LIMIT #{limit.to_i}" if (limit && limit > 0)
+      if key_values.empty?
+        return "SELECT * FROM #{keyspace}.#{table} #{limit} ;"
+      elsif key_values.count > primary_key.flatten.count
```
Comment (gbagnoli) [inline lib/corm/model.rb]: `self.primary_key_count` instead of `primary_key.flatten.count` ?

Comment (zeroed) [inline lib/corm/model.rb]: You're right.
:+1:

File: lib/corm/validations.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
+# encoding: utf-8
+
+module Corm
+  module Validations
+
+    def there_are_too_many_keys_requested?(key_values)
+      if key_values.keys.count > self.primary_key.flatten.count
+        return "You defined more find keys than the primary keys of the table!"
+      end
+      nil
```
Comment (gbagnoli) [inline lib/corm/validations.rb]: I'd prefer these functions used for validation to return `false`, instead of `nil`.

**Comment by gbagnoli** [comment]
@zeroed w.r.t. `Cassandra::Errors::IOError: Too many open files`: I don't get the error :eyes:

File: lib/corm/model.rb (hunk 9 above)
Comment (zeroed) [inline lib/corm/model.rb:210]: I think that it's in big part some kind of fancy readability... that kind of things that could became a bug...
:smile:

File: lib/corm/model.rb (hunk 10 above)
Comment (gbagnoli) [inline lib/corm/model.rb]: same thing in the validations file as well

File: lib/corm/validations.rb (hunk 11 above)
Comment (zeroed) [inline lib/corm/validations.rb]: POLA ... why not.
I agree...

File: lib/corm/validations.rb [hunk 12]
```diff
@@ -0,0 +1,56 @@
+# encoding: utf-8
+
+module Corm
+  module Validations
+
+    def there_are_too_many_keys_requested?(key_values)
+      if key_values.keys.count > self.primary_key_count
+        return "You defined more find keys than the primary keys of the table!"
```
Comment (zeroed) [inline lib/corm/validations.rb:8]: I just want to be clear that, right now, this error message is not used by the corresponding exception... not yet.

**Comment by zeroed** [comment]
I'll delete this branch when the #7 will be merged.
